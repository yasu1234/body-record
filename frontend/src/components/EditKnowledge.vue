<script setup>
import { ref, onMounted } from 'vue';
import axios from 'axios';
import { useRoute } from 'vue-router'
import Cookies from 'js-cookie';

import DropFile from './DropFile.vue'

const route = useRoute();

const title = ref("");
const knowledge = ref("");
const files = ref([]);
const imageUrls = ref([]);

onMounted(() => {
    getDetail();
});

function dateChange(event) {
    recordDate.value = event
}

function onFileChange(event) {
    files.value = [...event];
}

const getDetail = async () => {
    const id = route.params.id
    try {
        const res = await axios.get(import.meta.env.VITE_APP_API_BASE + `/api/v1/knowledges/${id}`, {
            headers: {
                'access-token' : Cookies.get('accessToken'),
                'client':Cookies.get('client'),
                'uid': Cookies.get('uid')
            }
        })
        console.log({ res })
        title.value = res.data.knowledge.title
        knowledge.value = res.data.knowledge.knowledge
        imageUrls.value = res.data.imageUrls
    } catch (error) {
        console.log({ error })
    }
}
</script>

<template>
    <div class="editor">
        <label class="itemLabel">タイトル</label>
        <input id="title" type="text" v-model="title">
        <textarea name="knowledge" rows="20" v-model="knowledge"></textarea>
    </div>
    <div class="thumbnail" v-for="item in imageUrls">
        <img :src="item" alt="">
    </div>
    <div class="relationImages">
        <p class="inputTitle">関連画像</p>
        <DropFile @change="onFileChange"/>
    </div>
    <div class="relationImages">
        <button class="registerButton" @click="registerRecord">登録する</button>
    </div>
</template>

<style>
.editor{
   padding: 30px;
 }

 input[type=text] {
  width: 100%;
  padding: 12px 12px;
  margin: 8px 0;
  box-sizing: border-box;
  border: 1px solid #ccc;
  border-radius: 4px;
}
.time-list {
  display: flex;
  justify-content: space-between;
  padding-left: 30px;
  padding-right: 30px;
}
.time-list .item {
  width: 50%;
  margin: 0;
  padding: 10px;
  box-sizing: border-box;
}
.time-list .item .inputTitle {
  margin: 5px 0 0;
  padding: 0;
  font-size: 16px;
}
.relationImages {
    padding: 20px;
}
.registerButton{
    background: #ffa500;
    color: white;
    font-size:16px;
    font-weight:bold;
}
.thumbnail {
    display: inline-block;
    height: 200px;
    margin-right: 15px;
    margin-bottom: 15px;
    padding-left: 20px;
}
.thumbnail img {
    height: 100%;
}
</style>