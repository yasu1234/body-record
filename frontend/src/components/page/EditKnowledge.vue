<script setup>
import { ref, onMounted } from 'vue';
import axios from 'axios';
import { useRoute } from 'vue-router'
import Cookies from 'js-cookie';

import DropFile from '../atom/DropFile.vue'
import Header from '../layout/Header.vue'

const route = useRoute();

const title = ref("");
const knowledge = ref("");
const files = ref([]);
const imageUrls = ref([]);
const knowledgeId = ref(null);

onMounted(() => {
    getDetail();
});

function dateChange(event) {
    recordDate.value = event
}

function onFileChange(event) {
    files.value = [...event];
}

const deleteImage = async (item) => {
    try {
        const res = await axios.delete(import.meta.env.VITE_APP_API_BASE + '/api/v1/knowledges/image', {
            params: {
                'id': knowledgeId.value,
                'image_id': item.id
            },
            headers: {
                'access-token' : Cookies.get('accessToken'),
                'client':Cookies.get('client'),
                'uid': Cookies.get('uid')
            }
        })
        console.log({ res })
    } catch (error) {
        console.log({ error })
    }
}

const getDetail = async () => {
    const id = route.params.id
    try {
        const res = await axios.get(import.meta.env.VITE_APP_API_BASE + `/api/v1/knowledges/${id}`, {
            headers: {
                'access-token' : Cookies.get('accessToken'),
                'client':Cookies.get('client'),
                'uid': Cookies.get('uid')
            }
        })
        knowledgeId.value = res.data.knowledge.id
        title.value = res.data.knowledge.title
        knowledge.value = res.data.knowledge.content
        imageUrls.value = res.data.imageUrls
    } catch (error) {
        console.log({ error })
    }
}

const edit = async () => {
    try {
        const formData = new FormData();
        formData.append('knowledge[title]', title.value);
        formData.append('knowledge[content]', knowledge.value);

        for (const file of files.value) {
            formData.append('images', file);
        }

        const res = await axios.patch(import.meta.env.VITE_APP_API_BASE + `/api/v1/knowledges/${knowledgeId.value}`, formData, {
            headers: {
                'Content-Type': 'multipart/form-data',
                'access-token' : Cookies.get('accessToken'),
                'client':Cookies.get('client'),
                'uid': Cookies.get('uid')
            }
        })
        knowledgeId.value = res.data.knowledge.id
        title.value = res.data.knowledge.title
        knowledge.value = res.data.knowledge.content
        imageUrls.value = res.data.imageUrls
    } catch (error) {
        console.log({ error })
    }
}
</script>

<template>
    <Header />
    <div class="editor">
        <label class="itemLabel">タイトル</label>
        <input id="title" type="text" v-model="title">
        <textarea name="knowledge" rows="20" v-model="knowledge"></textarea>
    </div>
    <div class="thumbnail-container">
    <div class="thumbnail" v-for="item in imageUrls">
      <div class="thumbnail-image">
        <img :src="item.url" alt="">
      </div>
      <div class="thumbnail-actions">
        <button class="delete-button" @click="deleteImage(item)">X</button>
      </div>
    </div>
  </div>
    <div class="relationImages">
        <p class="inputTitle">関連画像</p>
        <DropFile @change="onFileChange"/>
    </div>
    <div class="relationImages">
        <button class="registerButton" @click="edit">登録する</button>
    </div>
</template>

<style>
.editor{
   padding: 30px;
 }

 input[type=text] {
  width: 100%;
  padding: 12px 12px;
  margin: 8px 0;
  box-sizing: border-box;
  border: 1px solid #ccc;
  border-radius: 4px;
}
.time-list {
  display: flex;
  justify-content: space-between;
  padding-left: 30px;
  padding-right: 30px;
}
.time-list .item {
  width: 50%;
  margin: 0;
  padding: 10px;
  box-sizing: border-box;
}
.time-list .item .inputTitle {
  margin: 5px 0 0;
  padding: 0;
  font-size: 16px;
}
.relationImages {
    padding: 20px;
}
.registerButton{
    background: #ffa500;
    color: white;
    font-size:16px;
    font-weight:bold;
}
.thumbnail-container {
    display: flex;
    flex-wrap: wrap;
    justify-content: flex-start;
    padding: 20px;
}
.thumbnail {
    position: relative;
    display: inline-block;
    height: 200px;
    margin-right: 15px;
    margin-bottom: 15px;
    padding-left: 20px;
}
.thumbnail img {
    height: 100%;
}
.thumbnail-image {
    height: 100%;
}
.thumbnail-image img {
    height: 100%;
}
.thumbnail-actions {
    position: absolute;
    top: 5px;
    right: 5px;
}
.delete-button {
    background-color: rgba(255, 255, 255, 0.7);
    border: none;
    color: #000;
    padding: 5px 10px;
    text-align: center;
    text-decoration: none;
    display: inline-block;
    font-size: 12px;
    border-radius: 4px;
    cursor: pointer;
    border-radius: 50%
  }
</style>