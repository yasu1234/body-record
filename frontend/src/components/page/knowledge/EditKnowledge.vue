<template>
  <Header />
  <TabMenu />
  <Toast position="top-center" />
  <div class="p-7">
    <FloatLabel class="mt-5">
      <InputText v-model="title" class="input-width" />
      <label>タイトル</label>
    </FloatLabel>
    <div class="mt=2.5">
      <KnowledgeContentInput
        :knowledgeContent="knowledge"
        @content-edit="contentEdit"
      />
    </div>
    <div class="markdown-info">
      マークダウンを一部使用できます。詳しくは<a
        href="https://kumaapp.amebaownd.com/posts/54283163"
        target="_blank"
        rel="noopener noreferrer"
        class="info-link"
        >こちら</a
      >をご覧ください。<br>
      特に複数行の改行は使うことが多いと思うので必ずご確認ください
    </div>
  </div>
  <div v-if="imageUrls !== null && imageUrls.length !== 0">
    <p class="mt-5 ml-5">登録済みの画像</p>
    <div class="thumbnail-container">
      <div class="thumbnail" v-for="item in imageUrls">
        <div class="thumbnail-image">
          <img :src="item.url" alt="" />
        </div>
        <div class="thumbnail-actions">
          <Button
            label=""
            icon="pi pi-trash"
            v-tooltip="{ value: '画像削除' }"
            class="delete-button"
            @click="deleteImage(item)"
          />
        </div>
      </div>
    </div>
  </div>
  <div class="p-5">
    <p>関連画像(5枚まで登録できます)</p>
    <div class="file-input-container">
      <div v-for="i in 5">
        <DropFile @change="onFileChange" :index="i" class="mt-3" />
      </div>
    </div>
  </div>
  <div class="p-10 text-center">
    <button class="edit-knowledge-button" @click="edit">編集する</button>
  </div>
</template>

<script setup>
import { ref, onMounted } from "vue";
import { useRoute, useRouter } from "vue-router";
import { useToast } from "primevue/usetoast";
import { toastService } from "../../../js/toast.js";
import Toast from "primevue/toast";
import Button from "primevue/button";
import FloatLabel from "primevue/floatlabel";
import InputText from "primevue/inputtext";
import { axiosInstance, setupInterceptors } from "../../../js/axios.js";

import DropFile from "../../atom/DropFile.vue";
import Header from "../../layout/Header.vue";
import TabMenu from "../../layout/TabMenu.vue";
import KnowledgeContentInput from "../../layout/KnowledgeContentInput.vue";

const route = useRoute();
const router = useRouter();
const toast = useToast();
const toastNotifications = new toastService(toast);
setupInterceptors(router);

const title = ref("");
const knowledge = ref("");
const files = ref([...Array(5)]);
const imageUrls = ref([]);
const knowledgeId = ref(null);

onMounted(() => {
  getDetail();
});

const deleteImage = async (item) => {
  try {
    const res = await axiosInstance.delete("/api/v1/knowledge_images", {
      params: {
        id: knowledgeId.value,
        image_id: item.id,
      },
    });
    imageUrls.value = res.data.image_urls;
  } catch (error) {
    if (error.response == null) {
      toastNotifications.displayError("画像の削除に失敗しました", "");
      return;
    }

    let errorMessages = "";
    if (error.response.status === 422) {
      if (Array.isArray(error.response.data.errors)) {
        errorMessages += error.response.data.errors.join("\n");
      } else {
        errorMessages = error.response.data.errors;
      }
    } else if (error.response.status === 401) {
      errorMessages = "ログインしてください";
    }

    toastNotifications.displayError("画像の削除に失敗しました", errorMessages);
  }
};

const getDetail = async () => {
  const id = route.params.id;
  try {
    const res = await axiosInstance.get(`/api/v1/knowledges/${id}`);
    knowledgeId.value = res.data.knowledge.id;
    title.value = res.data.knowledge.title;
    knowledge.value = res.data.knowledge.content;
    imageUrls.value = res.data.knowledge.image_urls;
  } catch (error) {
    let errorMessage = "";
    if (error.response != null && error.response.status === 401) {
      errorMessage = "ログインしてください";
    }
    toastNotifications.displayError("登録データの取得に失敗しました", errorMessage);
  }
};

const edit = async () => {
  try {
    const formData = new FormData();
    formData.append("knowledge[title]", title.value);
    formData.append("knowledge[content]", knowledge.value);

    for (const file of files.value) {
      if (file != null) {
        formData.append("knowledge[images][]", file);
      }
    }

    const res = await axiosInstance.patch(
      `/api/v1/knowledges/${knowledgeId.value}`,
      formData,
      {
        headers: {
          "Content-Type": "multipart/form-data",
        },
      }
    );

    toastNotifications.displayInfo("編集しました", "");
    setTimeout(async () => {
      showKnowledgeDetail(res.data.knowledge);
    }, 3000);
  } catch (error) {
    if (error.response == null) {
      toastNotifications.displayError("記事の編集に失敗しました", "");
      return;
    }

    let errorMessage = "";

    if (error.response.status === 422) {
      if (Array.isArray(error.response.data.errors)) {
        errorMessage += error.response.data.errors.join("\n");
      } else {
        errorMessage = error.response.data.errors;
      }
    } else if (error.response.status === 401) {
      errorMessage = "ログインしてください";
    }

    toastNotifications.displayError(
      "記事の編集に失敗しました",
      errorMessage
    );
  }
};

const onFileChange = (event, index) => {
  files.value[index - 1] = event;
};

const contentEdit = (editContent) => {
  knowledge.value = editContent;
};

const showKnowledgeDetail = (item) => {
  router.push({ name: "KnowledgeDetail", params: { id: item.id } });
};
</script>

<style scoped>
.input-width {
  width: 100%;
  padding: 10px;
}
.edit-knowledge-button {
  font-size: 16px;
  font-weight: bold;
  padding: 10px 50px;
}
.thumbnail-container {
  display: flex;
  flex-wrap: wrap;
  justify-content: flex-start;
  padding: 20px;
}
.thumbnail {
  position: relative;
  display: inline-block;
  margin-right: 15px;
  margin-bottom: 15px;
  padding-left: 20px;
}
.thumbnail-image {
  height: 200px;
  width: 200px;
}
.thumbnail-image img {
  height: 200px;
  width: 200px;
}
.thumbnail-actions {
  position: absolute;
  top: 5px;
  right: 5px;
}
.delete-button {
  background-color: rgba(255, 255, 255, 0.7);
  border: none;
  color: #000;
  padding: 5px 10px;
  text-align: center;
  text-decoration: none;
  display: inline-block;
  font-size: 12px;
  border-radius: 4px;
  cursor: pointer;
  border-radius: 50%;
}
.markdown-info {
  font-size: 14px;
  color: #333;
}
.info-link {
  color: #ffa500;
  text-decoration: underline;
  cursor: pointer;
}
.info-link:hover {
  color: #ffa500;
}
</style>
