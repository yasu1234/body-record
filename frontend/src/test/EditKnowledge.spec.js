import { shallowMount } from "@vue/test-utils";
import EditKnowledge from "@/components/page/knowledge/EditKnowledge.vue";

jest.mock("axios");
jest.mock("vue-router", () => {
  const router = {
    push: jest.fn(),
    pathname: "/",
    route: "",
    asPath: "",
    query: {},
    events: {
      on: jest.fn(),
      off: jest.fn(),
      emit: jest.fn(),
    },
  };
  return {
    useRouter: jest.fn().mockReturnValue(router),
    useRoute: jest.fn().mockReturnValue({
      params: {
        id: 1,
      },
    }),
  };
});

jest.mock("primevue/usetoast", () => ({
  useToast: jest.fn(() => ({
    add: jest.fn(),
    removeAll: jest.fn(),
  })),
}));

jest.mock("@/js/axios.js", () => ({
  axiosInstance: {
    get: jest.fn(),
    interceptors: {
      request: { use: jest.fn() },
      response: { use: jest.fn() },
    },
  },
  setupInterceptors: jest.fn(),
}));

describe("EditKnowledge", () => {
  let wrapper;
  let mockedAxiosInstance;

  const mockKnowledge = {
    id: 1,
    title: "Test Knowledge",
    content: "This is a test knowledge",
    image_urls: [
      { id: 1, url: "https://example.com/image1.jpg" },
      { id: 2, url: "https://example.com/image2.jpg" },
    ],
  };

  beforeEach(() => {
    jest.mock("@/js/axios.js", () => ({
      axiosInstance: {
        get: jest.fn(),
        post: jest.fn(),
        delete: jest.fn(),
        interceptors: {
          request: { use: jest.fn() },
          response: { use: jest.fn() },
        },
      },
      setupInterceptors: jest.fn(),
    }));

    // モックされたaxiosInstanceを取得
    const axiosModule = require("@/js/axios.js");
    mockedAxiosInstance = axiosModule.axiosInstance;

    // onMounted内でaxios.getが呼ばれるため、このタイミングでモック設定
    mockedAxiosInstance.get.mockResolvedValue({ data: { knowledge: mockKnowledge } });

    wrapper = shallowMount(EditKnowledge, {
      global: {
        stubs: ["router-link", "router-view"],
      },
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("SnapShotテスト初期表示", () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
