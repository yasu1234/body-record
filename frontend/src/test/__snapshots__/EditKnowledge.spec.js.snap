// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditKnowledge SnapShotテスト初期表示 1`] = `
<div
  data-v-app=""
>
  
  <header-stub />
  <tab-menu-stub />
  <toast-stub
    position="top-center"
  />
  <div
    class="p-7"
  >
    <float-label-stub
      class="mt-5"
    />
    <div
      class="mt=2.5"
    >
      <knowledge-content-input-stub
        knowledgecontent="This is a test knowledge"
      />
    </div>
  </div>
  <div>
    <p
      class="mt-5 ml-5"
    >
      登録済みの画像
    </p>
    <div
      class="thumbnail-container"
    >
      
      <div
        class="thumbnail"
      >
        <div
          class="thumbnail-image"
        >
          <img
            alt=""
            src="https://example.com/image1.jpg"
          />
        </div>
        <div
          class="thumbnail-actions"
        >
          <button-stub
            class="delete-button"
            icon="pi pi-trash"
            label=""
          />
        </div>
      </div>
      <div
        class="thumbnail"
      >
        <div
          class="thumbnail-image"
        >
          <img
            alt=""
            src="https://example.com/image2.jpg"
          />
        </div>
        <div
          class="thumbnail-actions"
        >
          <button-stub
            class="delete-button"
            icon="pi pi-trash"
            label=""
          />
        </div>
      </div>
      
    </div>
  </div>
  <div
    class="p-5"
  >
    <p>
      関連画像(5枚まで登録できます)
    </p>
    <div
      class="file-input-container"
    >
      
      <div>
        <drop-file-stub
          class="mt-3"
          index="1"
        />
      </div>
      <div>
        <drop-file-stub
          class="mt-3"
          index="2"
        />
      </div>
      <div>
        <drop-file-stub
          class="mt-3"
          index="3"
        />
      </div>
      <div>
        <drop-file-stub
          class="mt-3"
          index="4"
        />
      </div>
      <div>
        <drop-file-stub
          class="mt-3"
          index="5"
        />
      </div>
      
    </div>
  </div>
  <div
    class="p-10 text-center"
  >
    <button
      class="edit-knowledge-button"
    >
      編集する
    </button>
  </div>
  
</div>
`;
